trieste/acquisition/function.py

```diff
@@ -509,6 +512,161 @@ def prepare_acquisition_function(
+        :return: The expected_hv_improvement function.
+        """
+        tf.debugging.assert_positive(len(dataset), message="Dataset must be populated.")
+        mean, _ = model.predict(dataset.query_points)
+
+        _pf = Pareto(mean)
+        _reference_pt = get_reference_point(_pf.front)
+        return expected_hv_improvement(model, _pf, _reference_pt)
+
+
+def expected_hv_improvement(
+    model: ProbabilisticModel,
+    pareto: Pareto,
+    reference_point: TensorType,
+) -> AcquisitionFunction:
+    r"""
+    expected Hyper-volume (HV) calculating using Eq. 44 of :cite:`yang2019efficient` paper.
+    The expected hypervolume improvement calculation in the non-dominated region
+    can be decomposed into sub-calculations based on each partitioned cell.
+    For easier calculation, this sub-calculation can be reformulated as a combination
+    of two generalized expected improvements, corresponding to Psi (Eq. 44) and Nu (Eq. 45)
+    function calculations, respectively.
+    Note:
+    1. Since in Trieste we do not assume the use of a certain non-dominated partition algorithm,
+       we do not assume the last dimension partitioned cell has only one (lower) bound
+       (which is used in the :cite:`yang2019efficient` paper). This is not as efficient as
+        the original paper, but is applicable to different non-dominated partition algorithm.
+    2. As the Psi and nu function in the original paper are defined for maximization problems,
+        we inverse our minimisation problem (to also be a maximisation), allowing use of the
+        original notation and equations.
+
+    :param model: The model of the objective function.
+    :param pareto: Pareto class
+    :param reference_point The reference point for calculating hypervolume
+    :return The expected_hv_improvement acquisition function modified for objective
+        minimisation. This function will raise :exc:`ValueError` or
+        :exc:`~tf.errors.InvalidArgumentError` if used with a batch size greater than one.
+    """
+
+    def acquisition(x: TensorType) -> TensorType:
+        tf.debugging.assert_shapes(
+            [(x, [..., 1, None])],
+            message="This acquisition function only supports batch sizes of one.",
+        )
+        normal = tfp.distributions.Normal(
+            loc=tf.zeros(shape=1, dtype=x.dtype), scale=tf.ones(shape=1, dtype=x.dtype)
+        )
+
+        def Psi(a: TensorType, b: TensorType, mean: TensorType, std: TensorType) -> TensorType:
+            """
+            Generic Expected Improvement defined at Eq. 19 in :cite:`yang2019efficient`,
+            calculate the expected improvement on reference a, integrating from b
+            :param a: shape [num_cells, out_dim] expected improvement reference point
+            :param b: shape [num_cells, out_dim] integral upper bounds
+            :param mean: shape [..., out_dim]
+            :param std: shape [..., out_dim]
+            """
```

Comment:
docstrings for inner functions will never reach the docs site (trieste don't use them elsewhere, indeed you haven't used them above). Just sth you might want to know. I'd not bother with them

Reply:
Okay, then it has been mostly removed.